cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/cluster/Diffs.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.controller.cluster.operations.cluster;
+
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.extensions.StatefulSet;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiPredicate;
+
+import static io.strimzi.controller.cluster.resources.AbstractCluster.containerEnvVars;
+
+class Diffs {
+    static boolean differingScale(StatefulSet a, StatefulSet b) {
+        return (a == null) != (b == null)
+                || !a.getSpec().getReplicas().equals(b.getSpec().getReplicas());
+    }
+
+    static <T extends HasMetadata> boolean differingLabels(T a, T b) {
+        return (a == null) != (b == null)
+                || !a.getMetadata().getLabels().equals(b.getMetadata().getLabels());
+    }
+
+    private static <T> boolean differingLists(List<T> a, List<T> b, BiPredicate<T, T> elementsDiffer) {
+        if ((a == null) != (b == null)) {
+            return true;
+        }
+        if (a == null) { // => b is null too
+            return false;
+        }
+        if (a.size() != b.size()) {
+            return true;
+        }
+        Iterator<T> ai = a.iterator();
+        Iterator<T> bi = b.iterator();
+        while (ai.hasNext()) {
+            if (elementsDiffer.test(ai.next(), bi.next())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    static boolean differingContainers(StatefulSet a, StatefulSet b) {
+        return (a == null) != (b == null)
+                || differingLists(a.getSpec().getTemplate().getSpec().getContainers(),
+                    b.getSpec().getTemplate().getSpec().getContainers(),
+                    Diffs::differingContainers);
+    }
+
+    static boolean differingContainers(Container a, Container b) {
+        return (a == null) != (b == null)
+                || !Objects.equals(a.getImage(), b.getImage())
+                || (a.getReadinessProbe() == null) != (b.getReadinessProbe() == null)
+                || !Objects.equals(a.getReadinessProbe().getInitialDelaySeconds(), b.getReadinessProbe().getInitialDelaySeconds())
+                || !Objects.equals(a.getReadinessProbe().getTimeoutSeconds(), b.getReadinessProbe().getTimeoutSeconds());
+    }
+
+    static boolean differingEnvironments(Container a, Container b, Set<String> keys) {
+        Map<String, String> aVars = containerEnvVars(a);
+        Map<String, String> bVars = containerEnvVars(b);
+        if (keys != null) {
+            aVars.keySet().retainAll(keys);
+            bVars.keySet().retainAll(keys);
+        }
+        return !aVars.equals(bVars);
+    }
+}
```

Comment:
Good question. The only thing we support changing about storage config is the deleteClaim flag. Any other change resulted in a warning message being logged during update of the SS. But from my reading of the code in `master` it would be possible to have modified the SS's claim 'by hand' and the CC would not have fixed it, because before we always took the SS from kubernetes and updated it in place when we patched. We now always patch it with what the SS should look like according to `KafkaCluter.generateStatefulSet()`, so any changes will get overwritten (such overwrites might not be immediately apparent, because we don't always do a rolling update, but they will get updated eventually).

So there are two potential problems here:

1. How can we tell if the user has changed the claim config if the SS has been changed directly?

2. Is the decision about rolling updates really detailed enough? Or should we compute the proper JSON-level diff (as fabric8 does) and use that to decide? Currently we can end up patching the SS but not doing a rolling update.

Thoughts?

Reply:
> Or should we compute the proper JSON-level diff (as fabric8 does) and use that to decide? 

To elaborate... fabric8 basically just does

```
JsonNode diff = JsonDiff.asJson(patchMapper().valueToTree(current), patchMapper().valueToTree(updated));
```

to compute a [JSON patch](http://erosb.github.io/post/json-patch-vs-merge-patch/#json-patch). We can call the same code to compute our own copy of the `diff` and see which `path`s had changed. If this was empty we wouldn't need to continue with the patching at all (better NOOP detection). If it included a path for the volume claim (`/spec/volumeClaimTemplates`) we could "reject" the change (log an error, change what's being updated, possibly also "correct" the cluster CM).

This would mean we had code to interpret [JSON paths](https://tools.ietf.org/html/rfc6902), instead of to diff two resources. Fabric8 provides some APIs for doing that.